blog/2020/05/invalidations.md

```diff
@@ -0,0 +1,518 @@
+knowing that only `f(::Int)` will be used: the call will be "hard-wired" into the code that Julia produces.
+You can see this using `@code_typed`:
+
+```
+julia> @code_typed applyf([100,200])
+CodeInfo(
+1 ─     Base.arrayref(true, container, 1)::Int64
+│       Base.arrayref(true, container, 2)::Int64
+└──     return 2
+) => Int64
+```
+
+The compiler itself knows that the answer will be 2, as long as the input array
+has elements indexable by 1 and 2. (Those `arrayref` statements enforce
+bounds-checking, and ensure that Julia will throw an appropriate error if you
+call `applyf([100])`.)
+
+If you pass a `Vector{Bool}`, it will compile `applyf` again, this time specializing it for `x::Bool`:
+
+```
+julia> @code_typed applyf([true,false])
+CodeInfo(
+1 ─     Base.arrayref(true, container, 1)::Bool
+│       Base.arrayref(true, container, 2)::Bool
+└──     return 4
+) => Int64
+```
+
+In this case, you can see that Julia knew those two `arrayref` statements would
+return a `Bool`, and since it knows the value of `f(::Bool)` it just went
+ahead and computed the result for you.
+
+At the end of these experiments, hidden away in Julia's "method cache" there will
+be two `MethodInstance`s of `applyf`, one specialized for `Vector{Int}` and the other specialized for `Vector{Bool}`.
+You don't normally see these, but Julia manages them for you; anytime you write
+code that calls `applyf`, it checks to see if this previous compilation work can be reused.
+
+For the purpose of this blog post, things start to get especially interesting if we try the following:
+```
+julia> c = Any[1, false];
+
+julia> applyf(c)
+3
+
+julia> @code_typed applyf(c)
+CodeInfo(
+1 ── %1  = Base.arrayref(true, container, 1)::Any
+│    %2  = (isa)(%1, Bool)::Bool
+└───       goto #3 if not %2
+2 ──       goto #6
+3 ── %5  = (isa)(%1, Int64)::Bool
+└───       goto #5 if not %5
+4 ──       goto #6
+5 ── %8  = Main.f(%1)::Int64
+└───       goto #6
+6 ┄─ %10 = φ (#2 => 2, #4 => 1, #5 => %8)::Int64
+│    %11 = Base.arrayref(true, container, 2)::Any
+│    %12 = (isa)(%11, Bool)::Bool
+└───       goto #8 if not %12
+7 ──       goto #11
+8 ── %15 = (isa)(%11, Int64)::Bool
+└───       goto #10 if not %15
+9 ──       goto #11
+10 ─ %18 = Main.f(%11)::Int64
+└───       goto #11
+11 ┄ %20 = φ (#7 => 2, #9 => 1, #10 => %18)::Int64
+│    %21 = Base.add_int(%10, %20)::Int64
+└───       return %21
+) => Int64
+```
+
+This may seem a lot more complicated, but the take-home message is actually
+quite simple. First, look at those `arrayref` statements: they are annotated
+`Any`, meaning that Julia can't predict in advance what they will return.
+Immediately after each reference, notice that there are two `isa` statements
+followed by a `ϕ`; all of this is essentially equivalent to
+
+```julia
+if isa(x, Bool)
+    value = 2
+elseif isa(x, Int)
+    value = 1
+else
+    value = f(x)::Int
+end
+```
+
+This is [union-splitting], a key optimization for performance in the face of uncertain types.
```

Comment:
While union splitting is indeed important I think this specific optimization (union splitting based on the number of methods on `f`) is not really a key optimization. I think most people would expect (and be fine) with a dynamic call here. After all, you're are just one based method overload away from getting `Any` anyway, so relying on this optimization feels pretty brittle.

The important case is imho when you infer that the input type is a union and then split on that.

Reply:
I rephrased it a bit, see if this is better.